src/wasm/wasm-type.cpp

```diff
@@ -2676,11 +2676,8 @@ bool RecGroupEquator::topLevelEq(HeapType a, HeapType b) const {
 }
 
 bool RecGroupEquator::eq(Type a, Type b) const {
-  if (a == b) {
-    return true;
-  }
```

Comment:
If `a == b`, but we're comparing them across two rec groups `A` and `B`, and `a` is a self-reference in `A` (i.e. a reference to a heap type in group `A`) but it is not a self-reference in `B`, then we know the rec groups have different shapes and are therefore different.

For example:

```
(rec (type $a (struct (ref $a))))
(rec (type $b (struct (ref $a))))
```

These rec groups look similar, but they are not the same because the first is self-referential and the second is not. `RecGroupEquator` would previously report that these rec groups are equal because it would see `$a` in the same location in both of them and return true because `$a == $a`.

Reply:
I've forgotten something basic, perhaps, but this function compares `Type`s. Wouldn't those two `$a`s refer to different types even though superficially they have the same name? Or is this before types are actually canonicalized? (If the latter then I'm confused as to how we can compare `Type` objects here.)

Reply:
No, the `$a` defined and used in the first group is the same type referred to as `$a` in the second group. Maybe what you're thinking of is that `$a` and `$b` are different types even though they superficially have the same structural definition?

Reply:
Right, sorry, the types are the same but the groups differ...